File: 01 Graph Theory/Graph Theory - disjoint_set_union.py
# DSU #
def findset(d):
    if d != p[d]: p[d] = findset(p[d])
    return p[d]

def union(a, b):
    a, b = findset(a), findset(b)
    if a != b:
        if s[a] < s[b]: a, b = b, a
        p[b] = a
        if s[a] == s[b]: s[a] += 1
            
def union(a, b):
    a, b = findset(a), findset(b)
    if a == b: return
    if s[a] >= s[b]:
        s[a] += s[b]
        p[b] = a
    else:
        s[b] += s[a]
        p[a] = b

File: 01 Graph Theory/Graph Theory - disjoint_set_union.cpp
class dsu{
public:
    int* parent;
    int* ranks;
    int n;
    mst(int N){
        parent = new int[N+1];
        ranks = new int[N+1];
        n = N;
        init();
    }
    void init(){
        for(int i=0; i<=n; i++){
            parent[i] = i;
            ranks[i] = 0;
        }
    }
    int find_parent(int node){
        if(parent[node] != node)
            return parent[node] = find_parent(parent[node]);
        return parent[node];
    }
    bool same_parent(int a, int b){
        return find_parent(a) == find_parent(b);
    }
    void union_parent(int a, int b){
        int parent_a = find_parent(a), parent_b = find_parent(b);
        if(parent_a == parent_b) return; // Same tree
        else if(ranks[parent_a]<ranks[parent_b]){ // B is deeper
            parent[parent_a] = parent_b;
        }
        else if(ranks[parent_b]<ranks[parent_a]){ // A is deeper
            parent[parent_b] = parent_a;
        }
        else{ // Same deep
            parent[parent_b] = parent_a;
            ranks[parent_a]++;
        }
    }
};
